les/server.go

```diff
@@ -140,179 +175,179 @@ func (s *LesServer) Stop() {
+// associated with the appropriate head hash by specific section index.
+func (s *LesServer) getCheckpoint(index uint64) (common.Hash, common.Hash, common.Hash) {
+	// convert last LES/2 section index back to LES/1 index for chtIndexer.SectionHead
+	latest := (index+1)*(light.CHTFrequencyClient/light.CHTFrequencyServer) - 1
 
-type requestCostStatsRlp []struct {
-	MsgCode uint64
-	Data    []byte
+	sectionHead := s.chtIndexer.SectionHead(latest)
+	chtRoot := light.GetChtRoot(s.protocolManager.chainDb, latest, sectionHead)
+	bloomTrieRoot := light.GetBloomTrieRoot(s.protocolManager.chainDb, index, sectionHead)
+	return sectionHead, chtRoot, bloomTrieRoot
 }
 
-var rcStatsKey = []byte("_requestCostStats")
-
-func newCostStats(db ethdb.Database) *requestCostStats {
-	stats := make(map[uint64]*linReg)
-	for _, code := range reqList {
-		stats[code] = &linReg{cnt: 100}
+// checkpointLoop starts a standalone goroutine to watch new checkpoint events and updates local's stable checkpoint.
+func (s *LesServer) checkpointLoop(checkpoint *light.TrustedCheckpoint) (err error) {
+	var (
+		eventCh      = make(chan *contract.ContractNewCheckpointEvent)
+		headCh       = make(chan core.ChainHeadEvent, SubscribeChainHeadEvent)
+		announcement = make(map[uint64]common.Hash)
+	)
+	eventSub, err := s.registrar.WatchNewCheckpointEvent(eventCh)
+	if err != nil {
+		return err
+	}
+	headSub := s.backend.SubscribeChainHeadEvent(headCh)
+	if headSub == nil {
+		eventSub.Unsubscribe()
+		return errors.New("subscribe head event failed")
 	}
 
-	if db != nil {
-		data, err := db.Get(rcStatsKey)
-		var statsRlp requestCostStatsRlp
-		if err == nil {
-			err = rlp.DecodeBytes(data, &statsRlp)
-		}
-		if err == nil {
-			for _, r := range statsRlp {
-				if stats[r.MsgCode] != nil {
-					if l := linRegFromBytes(r.Data); l != nil {
-						stats[r.MsgCode] = l
-					}
+	ticker := time.NewTicker(5 * time.Minute)
+	defer func() {
+		eventSub.Unsubscribe()
+		headSub.Unsubscribe()
+		ticker.Stop()
+	}()
+
+	for {
+		select {
+		case event := <-eventCh:
+			if event == nil {
+				// This should never happen.
+				log.Info("Ignore empty checkpoint event")
+				continue
+			}
+			// Note several events have same index may be received because of chain reorg and
+			// the modification of the latest checkpoint.
+			if checkpoint == nil || event.Index.Uint64() > checkpoint.SectionIdx {
+				log.Info("Receive new checkpoint event", "section", event.Index, "hash", common.Hash(event.CheckpointHash).Hex(),
+					"grantor", event.Grantor.Hex())
+				announcement[event.Index.Uint64()] = common.Hash(event.CheckpointHash)
+			}
+		case head := <-headCh:
+			number := head.Block.NumberU64()
+			if number < light.CheckpointConfirmations+light.CheckpointFrequency {
+				continue
+			}
+			idx := (number-light.CheckpointConfirmations)/light.CheckpointFrequency - 1
+			if checkpoint == nil || idx > checkpoint.SectionIdx {
+				hash, ok := announcement[idx]
```

Comment:
Here you are only noticing the match if the event filter found the announcement first and then the local indexer has calculated it. This is usually not the case because synced servers will finish processing more or less around the same time and the new checkpoint transaction also needs to be included in the chain and noticed by other nodes. You should detect the matching checkpoints it in either order.

Reply:
We will match and generate a stable checkpoint only when we receive a checkpoint announcement and the local indexer has calculated it. But the order doesn't matter here. Since whatever which one arrives first(In the normal case, local indexer should calculate it early than checkpoint announcement, but we keep the possibility if admin fetches the calculated unstable checkpoint from a `very fast node`, while our local node is `very slow node` which needs to take a long time to calculate unstable checkpoint), we will match the hash always. If announcement's hash is not equal with local one, we will keep match every block until find a valid one.